NETCONF RFC6241 and RFC6242 conformance fixes

jonasglass opened the pull request:
**NETCONF RFC6241 and RFC6242 conformance fixes**


__XML Version Conformance (RFC6241 Section 3)__
> All NETCONF messages MUST be well-formed XML ...
> A NETCONF message MAY begin with an XML declaration ...

Added XML declaration before every NETCONF message

__Hello Message Conformance (RFC6241 Section 8.1)__
> Each peer MUST send at least the base NETCONF capability, "urn:ietf:params:netconf:base:1.1"

Fixed wrong URN in hello message xmlns and capability (using 1.0 of course)

**RPC message conformance (RFC6241 Section 4.1)**
> The <rpc> element has a mandatory attribute "message-id", which is a
string chosen by the sender
 
Added the message-id element to RPCs. Also added XML NS to RPC as an attribute since it is present in all standards and most implementations.

**Framing Protocol Conformance (RFC6242 Section 4.1)**
The device prompt (`]]>]]>`) signals the end of a NETCONF message. In the current implementation the device prompt was only detected if it was followed by a \n since `readLine()` was used. This is not in conformance to the RFC. I changed it to reading character by character so that the end of a NETCONF message is also detected if the prompt is not being followed by a line break. 

**Using UTF8 encoding (RFC6241 Section 3)**
>All NETCONF messages MUST be well-formed XML, encoded in UTF-8

Changed encoding and decoding to force-use UTF-8 instead of system default.

**Extracting reused strings**
Extracted constants which have been used in multiple places and classes to a seperate class.

**Added debugging**
Added debug-logging upon receiving of a NETCONF messages. Currently it was only logged when sending a NETCONF message.

**Updated gitignore**
Added IntelliJ folders to git ignore.

peterjhill commented on src/main/java/net/juniper/netconf/Device.java:
```diff
@@ -7,11 +7,7 @@
-import com.jcraft.jsch.JSch;
-import com.jcraft.jsch.JSchException;
-import com.jcraft.jsch.Session;
+import com.jcraft.jsch.*;
```
I'd prefer not to use * imports.

peterjhill commented on src/main/java/net/juniper/netconf/Device.java:
```diff
@@ -143,23 +139,24 @@ public Device(
-        defaultCap.add("urn:ietf:params:xml:ns:netconf:base:1.0#confirmed-commit");
-        defaultCap.add("urn:ietf:params:xml:ns:netconf:base:1.0#validate");
-        defaultCap.add("urn:ietf:params:xml:ns:netconf:base:1.0#url?protocol=http,ftp,file");
+        defaultCap.add(NetconfConstants.URN_IETF_PARAMS_NETCONF_BASE_1_0);
```
love this.

peterjhill commented on src/main/java/net/juniper/netconf/Device.java:
```diff
@@ -776,21 +772,21 @@ public void commitFull() throws CommitException, IOException, SAXException {
-     *  services/&gt;&lt;/system&gt;&lt;/configuration/&gt;"
-     *  will load 'ftp' under the 'systems services' hierarchy.
+     *                   to be loaded. For example,
+     *                   " system {
```
The indents on the example are intentional. Please revert the example indentation.

peterjhill commented on src/main/java/net/juniper/netconf/NetconfConstants.java:
```diff
@@ -0,0 +1,31 @@
+package net.juniper.netconf;
```
this is great.

peterjhill commented:
P.S. The commit message is awesome!

jonasglass commented:
@peterjhill Thank you for your comments Peter! I changed the code to single imports and fixed the indention issues as well as bumping the minor version to 2.0.1

peterjhill commented:
Congrats @jonasglass on your first github PR!

jonasglass commented:
@peterjhill Thank you for the motivational hints and messages Peter! Nice to see such friendly community.
Next time I´ll try to squash the commits.
